### ISSUE TITLE
Dia causes crash in Unity 2017.3.0f3

### ISSUE THREAD

**Issue by spacehamster** [DaZombieKiller/TypeTreeTools#3]
Building type tree tools will cause the unity editor to crash while creating the Dia instance
```
0x00000000372B2204 (Mono JIT Code) (wrapper cominterop-invoke) Dia.IClassFactory:CreateInstance (object,System.Guid&,object&)
0x00000000372B24B2 (Mono JIT Code) [E:\TypeTreeTest\2017.3.0\Assembly-TypeTreeTools\Dia\IClassFactory.cs:25] Dia.IClassFactoryExtensions:CreateInstance<T_REF> (Dia.IClassFactory,object,T_REF&) 
0x00000000336B4367 (Mono JIT Code) [E:\TypeTreeTest\2017.3.0\Assembly-TypeTreeTools\Dia\DiaSourceFactory.cs:32] Dia.DiaSourceFactory:CreateInstance () 
0x000000003339F6D8 (Mono JIT Code) [E:\TypeTreeTest\2017.3.0\Assembly-TypeTreeTools\Unity\Core\PdbSession.cs:16] Unity.Core.PdbSession:.cctor ()
```
do you know away around this besides moving dia to a standalone commandline app?

Also, to get it to build on 2017.3, I had to replace `allReferences` in this line https://github.com/DaZombieKiller/TypeTreeTools/blob/40b77ef170a6cbf272d7516f5c94b079a45fd548/Assets/Editor/ProjectCompiler.cs#L66 with
```csharp
        static string[] GetDependencies(Assembly assembly)
        {
            // Older versions of unity don't include Mono/MonoBleedingEdge libraries 
            // (mscorlib.dll, System.dll, System.Core.dll and System.Xml.dll)
            // in assembly.allReferences
#if UNITY_2017
            var result = assembly.allReferences.ToList();
            var currentAssembly = System.Reflection.Assembly.Load(assembly.name);
            foreach (var name in currentAssembly.GetReferencedAssemblies())
            {
                var reference = System.Reflection.Assembly.Load(name);
                result.Add(reference.Location);
            }
            return result.Distinct().ToArray();
#else
        return assembly.allReferences;
#endif
        }
```

**Comment by DaZombieKiller** [comment]
> do you know away around this besides moving dia to a standalone commandline app?

Nothing else immediately comes to mind as an easy solution (other than maybe avoiding the `Marshal` APIs and using COM manually, or creating a native DLL that exposes parts of the API as a plain C interface). Earlier versions of this repository do have a standalone commandline app you can use as a base if you wish to add support for 2017 in your own fork though.

I'm planning to tidy up this project a bit later and make sure it works on older Unity versions, but right now it's only been tested with 2019.x and 2020.x (mostly 2020). I appreciate the heads up on `allReferences`, I'll look at incorporating that in a future commit.

**Comment by spacehamster** [comment]
I have been able to get Dia to work in 2017.3 using a native plugin that calls Dia directly, so i suspect it is a c# interop issue, but I haven't been able to diagnose the cause. 

Also if you are planning on porting to older version, the version you have now only works on 2019.3 and greater, it crashes on 2019.2 when calling `TypeTreeCache::GetTypeTree` (more specifically in `void __thiscall Release(TypeTreeShareableData *this)` during `TypeTree::operator=(class TypeTree const & __ptr64)` assignment). I am still looking into it, but it looks like the size of `TypeTree` is 144 bytes in 2019.2 and 24 bytes in 2019.3.

**Comment by DaZombieKiller** [comment]
> 
> 
> I have been able to get Dia to work in 2017.3 using a native plugin that calls Dia directly, so i suspect it is a c# interop issue, but I haven't been able to diagnose the cause.

Yeah, I believe it's a Mono runtime issue. Newer versions of Unity must have some changes related to COM interop.

> 
> Also if you are planning on porting to older version, the version you have now only works on 2019.3 and greater, it crashes on 2019.2 when calling `TypeTreeCache::GetTypeTree` (more specifically in `void __thiscall Release(TypeTreeShareableData *this)` during `TypeTree::operator=(class TypeTree const & __ptr64)` assignment). I am still looking into it, but it looks like the size of `TypeTree` is 144 bytes in 2019.2 and 24 bytes in 2019.3.

Yeah. `TypeTreeIterator` went through various changes throughout Unity versions as well. I haven't spent a lot of time looking into what the exact changes are though.

I've also been considering whether or not it's worth just figuring out the address of the internal type tree stored in a `SerializedObject` and calling `GetTypeTree` with that address rather than figuring out the full layout of the type tree for each version.

**Comment by spacehamster** [comment]
> I've also been considering whether or not it's worth just figuring out the address of the internal type tree stored in a SerializedObject and calling GetTypeTree with that address rather than figuring out the full layout of the type tree for each version.

I decided to look into this a little. this is a function `TypeTree * __thiscall GetTypeTreeForMultiEdit(SerializedObject *this,Object *param_1)` which looks like it may be possible to get the type tree from a SerializedObject without too much trouble, so the approach might work.

**Comment by DaZombieKiller** [comment]
Whoa, I never came across that function during my initial investigation. Do you know if it's present in older versions of the engine?

**Comment by spacehamster** [comment]
Its in 2017.3, i haven't checked any further back yet.

**Comment by spacehamster** [comment]
I checked some older versions, it appears in 4.7.2 but not in 4.0.0, so it was added sometime in between then. `GetPopupMenuData` might be useful if you want to target 4.0.0 or lower, but I doubt it is worth the effort.

**Comment by DaZombieKiller** [comment]
I believe `SerializedObject` was added as a public API in Unity 5.x, so I'm surprised it even exists in 4.x. If I ever intend to support 4.x, I'll probably implement a clone of `SerializedObject` to keep the number of necessary code changes and `#if`s down.

**Comment by DaZombieKiller** [comment]
It looks like `GetTypeTreeForMultiEdit` isn't present in 2020 (I checked with 2020.2.0a19 specifically), so it was removed at some point.

Edit: However, there is `public: bool __cdecl SerializedObject::GetManagedReferenceFromReferenceIterator(class SerializedProperty const & __ptr64,int & __ptr64,int & __ptr64,class TypeTree * __ptr64 & __ptr64) __ptr64` which may be useful.

**Comment by DaZombieKiller** [comment]
I've been looking at the full-size PDBs (available [here](http://symbolserver.unity3d.com/Unity_x64_mono.pdb/?C=M;O=A)) and there doesn't seem to be a difference between the 2019.2 and 2019.3 `TypeTree` class. `dynamic_array` doesn't appear to change either, although I might be looking at the wrong files (since I can only sort them by date, I'm not sure how to verify which version a PDB belongs to)

**Comment by spacehamster** [comment]
I was inspecting `TypeTreeCache::GetTypeTree` using ghidra.
Unity 2019.2.0 allocates 144 bytes for TypeTree
![image](https://user-images.githubusercontent.com/3821056/89775294-df619a80-db4a-11ea-8e71-44e8b054c485.png)
Unity 2019.3.0 allocates 24 bytes for TypeTree
![image](https://user-images.githubusercontent.com/3821056/89775349-f4d6c480-db4a-11ea-9e58-45cb4a34694e.png)
I was not aware there was a method of inspecting the struct/class layout from PDBs?

**Comment by DaZombieKiller** [comment]
> I was not aware there was a method of inspecting the struct/class layout from PDBs?

I was using the `pdb2hpp` tool from https://github.com/willglynn/pdb to inspect the PDB without needing to wait for Ghidra to process it. I also have some tools I wrote that use the DIA SDK to output a basic class layout.

**Comment by spacehamster** [comment]
Whats the syntax for `pdb2hpp`? running `cargo run --release --example pdb2hpp "C:\Program Files\Unity\Hub\Editor\2019.2.0f1\Editor\unity_x64.pdb" TypeTree` results in `sorry, class TypeTree was not found`

**Comment by DaZombieKiller** [comment]
~~`cargo run --release --example pdb2hpp -- pdb_path class_name`. You use `--` to separate the arguments to `cargo` and the arguments to `pdb2hpp`, just like how you would with `dotnet run`.~~

Oops, I completely misread that. The `unity_x64.pdb` file that comes with Unity has been trimmed, so it's missing a lot of type information. You need to grab one of the full PDBs from the link I posted earlier to get information on `TypeTree` and other classes.

I should also point out that when you download a PDB from that page, it'll be compressed (notice how the filename ends in `_`), most archival programs should be capable of decompressing them, I used 7-zip.

**Comment by spacehamster** [comment]
Thanks for the tip. It seems like you can download PDBs from the symbol server with 
```
cd C:\Program Files (x86)\Windows Kits\10\Debuggers\x64
symchk "C:\Program Files\Unity\Hub\Editor\2019.2.0f1\Editor\Unity.exe" /s srv*C:\DownstreamStore*http://symbolserver.unity3d.com/
```

**Comment by DaZombieKiller** [comment]
Yeah, although I'm pretty sure that will download one of the trimmed PDBs (from the `Unity_x64.pdb` directory instead of `Unity_x64_mono.pdb`)

**Comment by spacehamster** [comment]
I'm having trouble associating the `Unity_x64_mono.pdb`s with exact versions of the unity editor, but I have found a few versions of the 2018 version of the editor that use Unity_x64_mono.pdb, which I can find exact versions for.

It appears that the `UnityPlayer` PDBs are not stripped, so assuming there is not too much difference between the player and the editor, it may be more reliable to use PDBs from the player to generate struct layouts.

Here are the layouts from running pdb2hpp on `UnityPlayer_Win64_development_mono_x64.pdb` from the symbol server.

[2019.2.0_Player_TypeTree.txt](https://github.com/DaZombieKiller/TypeTreeTools/files/5054480/2019.2.0_Player_TypeTree.txt)

[2019.3.0_Player_TypeTree.txt](https://github.com/DaZombieKiller/TypeTreeTools/files/5054479/2019.3.0_Player_TypeTree.txt)

**Comment by DaZombieKiller** [comment]
There are a lot of things that are similar between the player and editor, but still a lot of differences. The type tree is stored differently between the two, which is the most important difference for this project. In the editor, `TypeTree` is just three `dynamic_array<T,0>` structs, but in the player it's a pointer to a `TypeTreeShareableData` and some pool information, as you can see in the dumps you posted. `TypeTreeShareableData` contains the three arrays that the editor normally has, but with additional information stored after.

If something like https://github.com/NationalSecurityAgency/ghidra/issues/2194 gets added to Ghidra, then we'd be able to use the full PDBs with any version of the editor, but at the moment the more likely solution would be to create a tool to generated merged PDBs. That's a massive undertaking though, considering the amount of stuff in the PDB format.

**Comment by spacehamster** [comment]
It may be easier to write a tool that exports type information from the full PDB to a custom format (say JSON for example), and then a ghidra java/python script that can import the custom type information before importing the trimmed PDB, rather then mess around with writing to the PDB itself.

**Comment by DaZombieKiller** [comment]
That's not a bad idea, I hadn't considered doing something like that. It'd definitely take less time than either of the solutions I just proposed.

**Comment by spacehamster** [comment]
> In the editor, TypeTree is just three dynamic_array<T,0> structs, but in the player it's a pointer to a TypeTreeShareableData and some pool information, as you can see in the dumps you posted. TypeTreeShareableData contains the three arrays that the editor normally has, but with additional information stored after.

I don't understand this comment, the TypeTree in the project has the the same layout as the UnityPlayer_Win64_development_mono_x64.pdb, not the Unity_x64_mono.pdb
https://github.com/DaZombieKiller/TypeTreeTools/blob/master/Assembly-TypeTreeTools/Unity/Core/TypeTree.cs
Decompiling the TypeTree Constructor also shows that it is assigning a pointer to a TypeTreeShareableData field.



2020.1.0f1 UnityEditor TypeTree Constructor [2020.1.0f1 UnityEditor TypeTree__TypeTree.txt](https://github.com/DaZombieKiller/TypeTreeTools/files/5060397/2020.1.0f1.UnityEditor.TypeTree__TypeTree.txt)
2020.1.0f UnityPlayer TypeTreeStruct [2020.1.0f UnityPlayer TypeTreeStruct.txt](https://github.com/DaZombieKiller/TypeTreeTools/files/5060395/2020.1.0f.UnityPlayer.TypeTreeStruct.txt)

From what i can see, the Unity_x64_mono.pdb struct layouts don't match the editor.

I decided to attach a debugger to see exactly where it crashed in unity 2019.2.0f1.
![image](https://user-images.githubusercontent.com/3821056/89968908-74be7500-dc98-11ea-8049-19eb968774f1.png)
![image](https://user-images.githubusercontent.com/3821056/89968993-a8999a80-dc98-11ea-821b-0a38d8679c6d.png)
![image](https://user-images.githubusercontent.com/3821056/89968950-8ef85300-dc98-11ea-95f2-1a67dd055567.png)


![image](https://user-images.githubusercontent.com/3821056/89969399-a126c100-dc99-11ea-80d1-fbada889681c.png)
![image](https://user-images.githubusercontent.com/3821056/89969356-85231f80-dc99-11ea-98bc-d1d3825c91a9.png)
![image](https://user-images.githubusercontent.com/3821056/89969336-750b4000-dc99-11ea-944e-1ee1fa175d54.png)


I think the main difference is 
2019.2.0
```c
// ??4TypeTree@@QEAAAEAV0@AEBV0@@Z

/* public: class TypeTree & __ptr64 __cdecl TypeTree::operator=(class TypeTree const & __ptr64)
   __ptr64 */

TypeTree * __thiscall operator_(TypeTree *this,TypeTree *param_1)

{
  TypeTreeShareableData *this_00;
  TypeTreeShareableData *pTVar1;
  
  this_00 = (TypeTreeShareableData *)(this + 8);
  if (*(TypeTreeShareableData **)this != this_00) {
    Release(*(TypeTreeShareableData **)this);
    *(undefined8 *)this = 0;
  }
  pTVar1 = *(TypeTreeShareableData **)param_1;
  if ((TypeTreeShareableData *)(param_1 + 8) == pTVar1) {
    operator_(this_00,(TypeTreeShareableData *)(param_1 + 8));
    *(TypeTreeShareableData **)this = this_00;
    return this;
  }
  *(TypeTreeShareableData **)this = pTVar1;
  LOCK();
  *(int *)(pTVar1 + 0x7c) = *(int *)(pTVar1 + 0x7c) + 1;
  return this;
}
```
2019.3.0
```
// ??4TypeTree@@QEAAAEAV0@AEBV0@@Z

/* public: class TypeTree & __ptr64 __cdecl TypeTree::operator=(class TypeTree const & __ptr64)
   __ptr64 */

TypeTree * __thiscall operator_(TypeTree *this,TypeTree *param_1)

{
  int *piVar1;
  longlong lVar2;
  
  if (*(TypeTreeShareableData **)this != (TypeTreeShareableData *)0x0) {
    Release(*(TypeTreeShareableData **)this);
    *(undefined8 *)this = 0;
  }
  lVar2 = *(longlong *)param_1;
  *(longlong *)this = lVar2;
  LOCK();
  piVar1 = (int *)(lVar2 + 0x7c);
  *piVar1 = *piVar1 + 1;
  lVar2 = *(longlong *)(param_1 + 8);
  if (lVar2 != 0) {
    LOCK();
    *(int *)(lVar2 + 0x28) = *(int *)(lVar2 + 0x28) + 1;
  }
  if ((this[0x10] == (TypeTree)0x0) && (*(Pool **)(this + 8) != (Pool *)0x0)) {
    Release(*(Pool **)(this + 8));
  }
  *(longlong *)(this + 8) = lVar2;
  this[0x10] = (TypeTree)0x0;
  return this;
}
```
is 
2019.2.0
```
  this_00 = (TypeTreeShareableData *)(this + 8);
  if (*(TypeTreeShareableData **)this != this_00) {
```
vs 
2019.3.0
```
  if (*(TypeTreeShareableData **)this != (TypeTreeShareableData *)0x0) {
```

Previous versions of the TypeTreeDumper allocated and constructed an instance of a TypeTree before calling GetTypeTree, could that be the issue?

**Comment by DaZombieKiller** [comment]
> I don't understand this comment, the TypeTree in the project has the the same layout as the UnityPlayer_Win64_development_mono_x64.pdb, not the Unity_x64_mono.pdb
> https://github.com/DaZombieKiller/TypeTreeTools/blob/master/Assembly-TypeTreeTools/Unity/Core/TypeTree.cs
> Decompiling the TypeTree Constructor also shows that it is assigning a pointer to a TypeTreeShareableData field.
> From what i can see, the Unity_x64_mono.pdb struct layouts don't match the editor.
> 

I just double-checked and you're correct. I haven't worked on this project for a while so I must've forgotten about the type tree differences between PDBs.

> Previous versions of the TypeTreeDumper allocated and constructed an instance of a TypeTree before calling GetTypeTree, could that be the issue?

I think that could possibly be it. I'm now wondering if `Release` is supposed to be called on `TypeTree`s that you receive from `TypeTreeCache` or not. I'd assume no, so when `Release` is re-implemented it'll probably need to be gated behind a version `#if`.

**Comment by spacehamster** [comment]
So I tried accessing Dia manually for 2017.
```csharp
Guid diaSourceClassGuid = new Guid("E6756135-1E65-4D17-8576-610761398C3C");
Type comType = Type.GetTypeFromCLSID(diaSourceClassGuid);
var obj = Activator.CreateInstance(comType);
```
With .net 3.5, the following error occurs.
```
NotImplementedException: The requested feature is not implemented.
System.Type.GetTypeFromCLSID (Guid clsid) (at /Users/builduser/buildslave/mono/build/mcs/class/corlib/System/Type.cs:526)
```
With .NET 4.6, the following error occurs
```
MethodAccessException: Method `System.__ComObject:.ctor ()' is inaccessible from method `System.__ComObject:.ctor ()'

<0x00000 + 0x00000> <unknown method>
System.__ComObject..ctor () (at <852ce84dd4974a728a855d74d9e1a5bc>:0)
(wrapper cominterop-invoke) System.__ComObject:.ctor ()
System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) (at <cfc149f8218b496788d8493c87de777a>:0)
Rethrow as TargetInvocationException: Exception has been thrown by the target of an invocation.
System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) (at <cfc149f8218b496788d8493c87de777a>:0)
System.RuntimeType.CreateInstanceMono (System.Boolean nonPublic) (at <cfc149f8218b496788d8493c87de777a>:0)
System.RuntimeType.CreateInstanceSlow (System.Boolean publicOnly, System.Boolean skipCheckThis, System.Boolean fillCache, System.Threading.StackCrawlMark& stackMark) (at <cfc149f8218b496788d8493c87de777a>:0)
System.RuntimeType.CreateInstanceDefaultCtor (System.Boolean publicOnly, System.Boolean skipCheckThis, System.Boolean fillCache, System.Threading.StackCrawlMark& stackMark) (at <cfc149f8218b496788d8493c87de777a>:0)
System.Activator.CreateInstance (System.Type type, System.Boolean nonPublic) (at <cfc149f8218b496788d8493c87de777a>:0)
System.Activator.CreateInstance (System.Type type) (at <cfc149f8218b496788d8493c87de777a>:0)
```

I get the feeling the only options are a native plugin wrapper or a command line wrapper. With that in mind, I wonder if it might be easier to do all of it in C++ to save converting the rest of the project .NET 3.5 just for 2017 and lower?

**Comment by DaZombieKiller** [comment]
> I get the feeling the only options are a native plugin wrapper or a command line wrapper. With that in mind, I wonder if it might be easier to do all of it in C++ to save converting the rest of the project .NET 3.5 just for 2017 and lower?

Yeah, a native plugin sounds like the way to go. An alternative could be a separate .NET application that simply hooks into the Unity process and retrieves the type tree information that way. That would be much easier to automate as well.

**Comment by spacehamster** [comment]
A seperate application would make calling functions and reading memory more difficult right? It is worth looking into.
I found
```
    [DllImport("kernel32.dll")]
    public static extern bool ReadProcessMemory(int hProcess,
    Int64 lpBaseAddress, byte[] lpBuffer, int dwSize, ref int lpNumberOfBytesRead);
```
for reading memory, but i'm not sure how you would call functions?

**Comment by DaZombieKiller** [comment]
It would require some experimentation with the `CreateRemoteThread` Windows API to inject a DLL, which would be used to call functions in the executable. (It might actually be easier to use EasyHook/CoreHook for that)

**Comment by DaZombieKiller** [comment]
I did some experimenting with the idea, and something like this does work: https://gist.github.com/DaZombieKiller/f5ecefa6a4fe7c66fd6afa6da7b5e5cb

**Comment by DaZombieKiller** [comment]
I've uploaded a super experimental prototype [here](https://github.com/DaZombieKiller/TypeTreeDumper), it just outputs the names of all types.

**Comment by spacehamster** [comment]
I tried it out on 2017 because I was having trouble getting managed Dia to run there. With some tweaking (adding try catch blocks and finally calling Environment.Exit(0), changing some symbols that don't exist in 2017), it works really well. It seems like it will be able to run modern versions of .NET with any version of Unity? 

In which case, it seems like a much better approach then the plan of a c++ plugin for < 2018 and a managed plugin for >= 2018.

**Comment by DaZombieKiller** [comment]
> It seems like it will be able to run modern versions of .NET with any version of Unity?

Yeah, since it runs the code independent of the Mono runtime included with Unity. It's also possible to use .NET Core if it's ported to CoreHook instead of EasyHook.
